### ISSUE TITLE
When using SignalR & EF entities are always detached

### ISSUE THREAD

**Issue by AllanMichaelsen** [aspnet/AspNetCore#18025]
I'm unsure if this belongs here or in EF, but I'll start here.

When I retrieve the entity the state is unchanged, but after round tripping to my application via SignalR my entity is always detached and that results in an the following scenatios:

- If I try to attach the entity it fails saying it's already tracked
- If I try to update the entity it fails with the same error

If I save changes nothing is persisted because the entity is detached and not tracked :-(

Any Ideas?

When I get my entity with this it's unchanged
`    public async Task<DtoContact> GetById(long contactId)
    {
      var contact = await GetFirstOrDefault<long, DtoContact>(x => x.ContactId == contactId, includeFlag: FlagContactInclude.CultureInfo | FlagContactInclude.ModifiedByUnit | FlagContactInclude.ModifiedByUser, includeActive: true, includeArchived: true);
      var state = context.Entry(contact)
       .State;
      return contact;
    }
`

When I return to this method to save the state is detached
`    public async Task<DtoContact> Upsert(DtoContact contact)
    {
      if (contact == null)
      {
        throw new ArgumentNullException(nameof(contact));
      }

      if (contact.ModifiedByUnit is null)
      {
        throw new ArgumentNullException(nameof(contact.ModifiedByUnit));
      }

      if (contact.ModifiedByUser is null)
      {
        throw new ArgumentNullException(nameof(contact.ModifiedByUser));
      }

      try
      {
        var state = context.Entry(contact)
         .State;
        if (state == EntityState.Added)
        {
          await context.Contact.AddAsync(contact);
        }
        else if (state == EntityState.Detached)
        {
          context.Contact.Attach(contact);
        }
        else if (state == EntityState.Modified)
        {
          context.Contact.Update(contact);
        }

        context.SaveChanges();

        return contact;
      }
      catch (Exception ex)
      {
        log.Error("Error upserting contact", ex);
        throw;
      }
    }
`

I'm out of ideas...

**Comment by davidfowl** [comment]
Yes they are always detached by design. You hub is completely stateless. Every invocation is a new hub instance created in a new scope for that invocation much like an http request.

**Comment by AllanMichaelsen** [comment]
So I should fetch from EF without tracking and the always attach on update calls?

**Comment by davidfowl** [comment]
> So I should fetch from EF without tracking and the always attach on update calls?

That's right.

**Comment by AllanMichaelsen** [comment]
I'll give it a go... Thanks..

**Comment by AllanMichaelsen** [comment]
So I'm using .AsNoTracking when getting my elements as described above. However I still get the error, but on "Child" elements.
The contact has a ModifiedByUser set to a default user, the contact also has a ModifiedByUnit which in turn also has a ModifiedByUser set to the same default user.
So when I attach the contact to the context it starts tracking the default user with the contacts ModifiedByUser and the tries again with the ModifiedByUser on the ModifiedByUnit and therefor it says already tracked. :-(
Any ideas?

**Comment by davidfowl** [comment]
@AllanMichaelsen did you read this https://docs.microsoft.com/en-us/ef/core/saving/disconnected-entities?

**Comment by AllanMichaelsen** [comment]
Yes and that works for modified elements but not when using `context.Contact.AddAsync(contact);`. In that case it still throws the error. 
I tried that previously but was looking for a general fix in both scenarios.
I have now modified my code to use `context.Entry(existingContact).CurrentValues.SetValues(contact);` for update, but I have no solution for adding new items.

**Comment by AllanMichaelsen** [comment]
This  seems to work, but it doesn't seem very elegant...

`        if (existingContact is null)
        {
          existingContact = new DtoContact();
          context.Entry(existingContact)
           .CurrentValues.SetValues(contact);
          existingContact = (await context.Contact.AddAsync(existingContact)).Entity;
        }
        else
        {
          context.Entry(existingContact)
           .CurrentValues.SetValues(contact);
        }
`

Update scenario: Getting the existing contact with tracking so the update part works.
New scenarion: making a new contact, settings the new values on the item and then adding it works.

But can it really be true that i nee the two first steps in the new scenario? Shouldn't I be able to just add the contact?

**Comment by davidfowl** [comment]
This is really an EF question with respect to disconnected graphs cc @ajcvickers @roji
